### ISSUE TITLE
jsonencode doesn't support nested structures => * jsonencode: map values must be strings

### ISSUE THREAD

**Issue by dennybaa** [hashicorp/terraform#9112]
Subj

```
variable "foo" {
    default = {
        hello = {
            world = "hello"
            man = "bar"
            john = {
                id = 1
            }
        }
    }
}

data "null_data_source" "json" {
    inputs {
        body = "${jsonencode(var.foo)}"
    }
}
```

**Comment by s-maj** [comment]
This would be extremely beneficial for managing ECS tasks. Instead of using templates or premade json files, it would be possible to create tasks straight in HCL.

**Comment by tamsky** [comment]
`terraform output -json` has supported nested structures since forever (afaict).
seems very odd that `jsonencode()` is not already using the same json function.

**Comment by onnimonni** [comment]
I bumped into this too but after updating `terraform 0.9.11 -> 0.11.7` this started to work.

This issue can be closed now.

**Comment by apparentlymart** [comment]
In Terraform v0.12.0-alpha1 we now have a fully-general `jsonencode` function that will accept values of any type.

I adapted the example given in the opening comment as a test:

```hcl
variable "foo" {
  default = {
    hello = {
      world = "hello"
      man   = "bar"
      john = {
        id = 1
      }
    }
  }
}

output "example" {
  value = jsonencode(var.foo)
}
```

```
$ terraform apply

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.

Outputs:

example = {"hello":{"john":{"id":1},"man":"bar","world":"hello"}}
```

Full details on how the various different Terraform language types are mapped to JSON can be found on [the `jsonencode` function's new documentation page](https://github.com/hashicorp/terraform/blob/c7c6f6efba23480f9553220788a7c9bf34313487/website/docs/configuration/functions/jsonencode.html.md), which will be published on the main website as part of the v0.12.0 final release.

An earlier release relaxed the rules for this function to allow for lists and maps of strings in addition to just strings, but the v0.12 release generalizes this completely to all of the types, including the new object type introduced with v0.12 that allows the objects representing entire resources to be JSON encoded if desired.
